packages/miniflare/src/shared/dev-registry.worker.ts

```diff
@@ -0,0 +1,349 @@
+					// There might be no HOST header when proxying a fetch request made over service binding
+					//  e.g. env.MY_WORKER.fetch("https://example.com")
+					requireHostHeader: false,
+				});
+
+				server.on("request", this.handleRequest.bind(this));
+				server.on("connect", this.handleConnect.bind(this));
+				server.listen(0, "127.0.0.1", () => {
+					const address = server.address();
+
+					if (!address || typeof address === "string") {
+						reject(new Error("Failed to get server address"));
+						return;
+					}
+
+					// Notify parent that server is ready with the port
+					this.messagePort.postMessage({
+						type: "ready",
+						address: `${address.address}:${address.port}`,
+					});
+
+					resolve();
+				});
+				server.on("error", reject);
+			});
+		} catch (error) {
+			this.messagePort.postMessage({
+				type: "error",
+				error: error instanceof Error ? error.message : String(error),
+			});
+		}
+	}
+
+	private handleRequest(
+		req: http.IncomingMessage,
+		res: http.ServerResponse
+	): void {
+		const serviceProxyTarget = extractServiceFetchProxyTarget(req);
+		if (serviceProxyTarget) {
+			const address = this.getServiceAddress(
+				serviceProxyTarget.service,
+				serviceProxyTarget.entrypoint
+			);
+
+			this.handleProxy(req, res, address);
+			return;
+		}
+
+		const doProxyTarget = extractDoFetchProxyTarget(req);
+		if (doProxyTarget) {
+			const address = this.getDurableObjectAddress(
+				doProxyTarget.scriptName,
+				doProxyTarget.className
+			);
+
+			if (!address) {
+				res.writeHead(503);
+				res.end("Service Unavailable");
+				return;
+			}
+
+			this.handleProxy(req, res, address);
+			return;
+		}
+
+		// If no valid target found, return 404
+		res.writeHead(404);
+		res.end("Not Found");
+	}
+
+	private async handleConnect(
+		req: http.IncomingMessage,
+		clientSocket: Duplex,
+		head: Buffer
+	) {
+		try {
+			const connectHost = req.url;
+			const [serviceName, entrypoint] = connectHost?.split(":") ?? [];
+			const address = this.getServiceAddress(serviceName, entrypoint);
+			const serverSocket = net.connect(address.port, address.host, () => {
+				serverSocket.write(`CONNECT ${HOST_CAPNP_CONNECT} HTTP/1.1\r\n\r\n`);
+
+				// Push along any buffered bytes
+				if (head && head.length) {
+					serverSocket.write(head);
+				}
+
+				serverSocket.pipe(clientSocket);
+				clientSocket.pipe(serverSocket);
+			});
+
+			// Errors on either side
+			serverSocket.on("error", (error) => {
+				log.error(error);
+				clientSocket.end();
+			});
+			clientSocket.on("error", () => serverSocket.end());
+			// Close the tunnel if the service is updated
+			// This make sure workerd will re-connect to the latest address
+			this.subscribe(serviceName, () => {
+				log.debug(`Closing tunnel as service "${serviceName}" was updated`);
+				clientSocket.end();
+			});
+		} catch (e) {
+			log.error(e instanceof Error ? e : new Error(`${e}`));
+			clientSocket.end();
+		}
+	}
```

Comment:
This was done in [handleLoopbackConnect](https://github.com/cloudflare/workers-sdk/pull/10142/files#r2252263534) before moving to the proxy server.